### ISSUE TITLE
Wrong gradient when clipping out of bounds indexing inside vmap

### ISSUE THREAD

**Issue by FFroehlich** [jax-ml/jax#25878]
### Description

The code below computes the gradient of the function

```
f(p,iy) = p.at[iy]
```

with `p=[1.0]` and compares result when using different automatic differentiation modes (`jax.grad`, `jax.jacrev`, `jax.jacfwd`), using `jax.vmap` or `jax.lax.map` and in bound (`iy=0`) and out of bound indexing (`iy=1`). 

Unexpectedly, the combinations `grad+vmap+out_of_bound` and `jacrev+vmap+out_of_bound` incorrectly produce all zero gradients. 

I would understand this kind of behavior for `at[iy].get(mode='promise_in_bounds')`, as the user does not deliver on the in bounds promise, but for any other mode this seems like a bug.

Code:
```python
import jax.numpy as jnp
from jax import grad, jacfwd, jacrev, vmap
from jax.lax import map

out_of_bound=jnp.ones(1).astype(jnp.int_)
in_bound=jnp.zeros(1).astype(jnp.int_)

def inner_fun(p: jnp.ndarray, iy: jnp.ndarray):
    return p.at[iy].get(mode="clip")

def outer_fun_vmap(p: jnp.ndarray, iy: jnp.ndarray):
    return vmap(inner_fun, in_axes=(None, 0))(p, iy)[0]

def outer_fun_map(p: jnp.ndarray, iy: jnp.ndarray):
    return map(lambda _iy: inner_fun(p, _iy), iy)[0]

for var_access, ix in (
    ('out_of_bound', out_of_bound),
    ('in_bound', in_bound),
):
    p = jnp.array([1.0])
    print(f'grad+vmap+{var_access}: {grad(outer_fun_vmap)(p, ix)}')
    print(f'jacrev+vmap+{var_access}: {jacrev(outer_fun_vmap)(p, ix)}')
    print(f'jacfwd+vmap+{var_access}: {jacfwd(outer_fun_vmap)(p, ix)}')

    print(f'grad+map+{var_access}: {grad(outer_fun_map)(p, ix)}')
    print(f'jacrev+map+{var_access}: {jacrev(outer_fun_map)(p, ix)}')
    print(f'jacfwd+map+{var_access}: {jacfwd(outer_fun_map)(p, ix)}')
```

Output
```
grad+vmap+out_of_bound: [0.]
jacrev+vmap+out_of_bound: [0.]
jacfwd+vmap+out_of_bound: [1.]
grad+map+out_of_bound: [1.]
jacrev+map+out_of_bound: [1.]
jacfwd+map+out_of_bound: [1.]
grad+vmap+in_bound: [1.]
jacrev+vmap+in_bound: [1.]
jacfwd+vmap+in_bound: [1.]
grad+map+in_bound: [1.]
jacrev+map+in_bound: [1.]
jacfwd+map+in_bound: [1.]
```

### System info (python version, jaxlib version, accelerator, etc.)

```
jax:    0.4.38
jaxlib: 0.4.38
numpy:  2.0.2
python: 3.12.7 (main, Oct  1 2024, 02:05:46) [Clang 15.0.0 (clang-1500.3.9.4)]
device info: cpu-1, 1 local devices"
process_count: 1
platform: uname_result(system='Darwin', node='LH7X112H6J', release='24.2.0', version='Darwin Kernel Version 24.2.0: Fri Dec  6 18:51:28 PST 2024; root:xnu-11215.61.5~2/RELEASE_ARM64_T8112', machine='arm64')
```

**Comment by jakevdp** [comment]
Thanks for the report! I think #25880 should fix the issue.
